wolfcrypt/src/asn.c

```diff
@@ -4666,6 +4666,156 @@ int wc_DhKeyDecode(const byte* input, word32* inOutIdx, DhKey* key, word32 inSz)
     return ret;
 }
 
+#ifdef WOLFSSL_DH_EXTRA
+
+/* Export DH Key (private or public) */
+int wc_DhKeyToDer(DhKey* key, byte* output, word32* outSz, int exportPriv)
+{
+    int ret, privSz = 0, pubSz = 0, keySz;
+    word32 idx, total;
+
+    if (key == NULL || outSz == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* determine size */
+    if (exportPriv) {
+        /* octect string: priv */
+        privSz = SetASNIntMP(&key->priv, -1, NULL);
+        idx = 1 + SetLength(privSz, NULL) + privSz; /* +1 for ASN_OCTET_STRING */
+    }
+    else {
+        /* bit string: public */
+        pubSz = SetASNIntMP(&key->pub, -1, NULL);
+        idx = 1 + SetLength(pubSz, NULL) + pubSz; /* +1 for ASN_BIT_STRING */
```

Comment:
BIT_STRING has an extra byte to indicate the number of unused bits in last octet.
Use SetBitString(pubSz, 0, NULL) to get length if possible.